src/orange/surf/Involute.hh

```diff
@@ -0,0 +1,347 @@
+ */
+CELER_FUNCTION SignedSense Involute::calc_sense(Real3 const& pos) const
+{
+    double const pi = 3.14159265358979323846;
+
+    /*
+     * Define tolerances.
+     */
+    real_type const tol = 1e-7;
+
+    real_type const x = pos[0] - origin_[0];
+    real_type const y = pos[1] - origin_[1];
+
+    /*
+     * Calculate distance to origin and obtain t value for disance.
+     */
+    real_type const rxy2 = ipow<2>(x) + ipow<2>(y);
+    real_type const tPoint2 = (rxy2 / ipow<2>(r_b_)) - 1;
+    real_type const tPoint = sqrt(tPoint2);
+
+    /*
+     * Check if Point is on involute.
+     */
+    real_type const angle = tPoint + a_;
+    real_type const xInv = r_b_ * (std::cos(angle) + tPoint * std::sin(angle));
+    real_type const yInv = r_b_ * (std::sin(angle) - tPoint * std::cos(angle));
+
+    if (abs(x - xInv) < tol && abs(y - yInv) < tol)
+    {
+        return SignedSense::on;
+    }
+
+    /*
+     * Check if point is in defined bounds.
+     */
+    if (abs(tPoint2) < ipow<2>(tmin_) - tol)
+    {
+        return SignedSense::outside;
+    }
+    if (abs(tPoint2) > ipow<2>(tmax_) + tol)
+    {
+        return SignedSense::outside;
+    }
+
+    /*
+     * Check if point is inside involute
+     */
+
+    // Calculate tangents
+    real_type a = rxy2;
+    real_type b;
+    real_type c;
+    real_type xa, xb;
+    real_type ya, yb, yc, yd;
+    real_type yalpha, ybeta;
+
+    b = -2 * ipow<2>(r_b_) * x;
+    c = ipow<4>(r_b_) - ipow<2>(y) * ipow<2>(r_b_);
+
+    xa = (-b + sqrt(ipow<2>(b) - 4 * a * c)) / (2 * a);
+    xb = (-b - sqrt(ipow<2>(b) - 4 * a * c)) / (2 * a);
+
+    ya = sqrt(ipow<2>(r_b_) - ipow<2>(xa));
+    yb = -sqrt(ipow<2>(r_b_) - ipow<2>(xa));
+    yc = sqrt(ipow<2>(r_b_) - ipow<2>(xb));
+    yd = -sqrt(ipow<2>(r_b_) - ipow<2>(xb));
+
+    b = -2 * ipow<2>(r_b_) * y;
+    c = ipow<4>(r_b_) - ipow<2>(x) * ipow<2>(r_b_);
+
+    yalpha = (-b + sqrt(ipow<2>(b) - 4 * a * c)) / (2 * a);
+    ybeta = (-b - sqrt(ipow<2>(b) - 4 * a * c)) / (2 * a);
+
+    Array<real_type, 2> point1;
+    Array<real_type, 2> point2;
+
+    // First tangent
+    if (abs(ya - yalpha) <= tol)
+    {
+        point1 = {xa, ya};
+    }
+    else
+    {
+        point1 = {xb, yc};
+    }
+    // Second tangent
+    if (abs(yb - ybeta) <= tol)
+    {
+        point2 = {xa, yb};
+    }
+    else
+    {
+        point2 = {xb, yd};
+    }
+
+    // Determine which tangent
+    Array<real_type, 2> point;
+    Array<real_type, 2> norm = {-y, x};
+    real_type dot = point1[0] * norm[0] + point1[1] * norm[1];
+    if (((0 < dot) - (dot < 0)) == sign_)
+    {
+        point = point1;
+    }
+    else
+    {
+        point = point2;
+    }
+
+    // Calculate angle of tangent
+    real_type theta = std::acos(
+        point[0] / sqrt(point[0] * point[0] + point[1] * point[1]));
+    if (point[1] < 0)
+    {
+        theta = (pi - theta) + pi;
+    }
+    real_type a1 = theta - tPoint;
+    if (abs(theta) < abs(tmax_ + a_) + tol && a1 >= a_ - tol)
```

Comment:
Always use `std::fabs` instead of `abs`, since the latter has a flaw in C/C++ where including the wrong header files may result in accidentally truncating to `int` before taking the absolute value.